### PR TITLE
Limited-Range Integer Number Datatypes

### PR DESCRIPTION
**Pull request by kaimal11**

### CODE REVIEW

**Comment by bigerl** [comment]
What is the matter with the `tests_Long` failing with all versions of clang?

see https://github.com/rdf4cpp/rdf4cpp/runs/6956113826?check_suite_focus=true#step:9:34

File: tests/datatype/tests_NegativeInteger.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    CHECK(lit2 != lit3);
+    CHECK(lit1 == lit4);
+    CHECK(lit2 == lit5);
+
+    // suppress warnings regarding attribute ‘nodiscard’
+    Literal no_discard_dummy;
+
+    CHECK_THROWS_WITH_AS(no_discard_dummy = Literal("a23dg", type_iri), "XSD Parsing Error", std::runtime_error);
+
+    CHECK_THROWS_WITH_AS(no_discard_dummy = Literal("0", type_iri), "XSD Parsing Error", std::runtime_error);
```
Comment (nkaralis) [inline tests/datatype/tests_NegativeInteger.cpp:54]: You could also add a test involving a positive number (>0).

File: src/rdf4cpp/rdf/datatypes/xsd/UnsignedByte.hpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+ * Name of the datatype. This is kept so that we won't need to type it over and over again.
+ */
+constexpr static registry::ConstexprString xsd_unsignedByte{"http://www.w3.org/2001/XMLSchema#unsignedByte"};
+
+/**
+ * Defines the mapping between the LiteralDatatype IRI and the C++ datatype.
+ */
+template<>
+struct DatatypeMapping<xsd_unsignedByte> {
+    using cpp_datatype = u_int8_t;
```
Comment (nkaralis) [inline src/rdf4cpp/rdf/datatypes/xsd/UnsignedByte.hpp]: uint8_t

File: src/rdf4cpp/rdf/datatypes/xsd/UnsignedByte.hpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * Specialisation of to_string template function.
+ */
+template<>
+inline std::string LiteralDatatypeImpl<xsd_unsignedByte>::to_string(const cpp_type &value) {
+
+    /**
+     * If not converted to int16_t, conversion to string generates a byte representation '\x01' which doest match the lexical representation.
+     */
+    int conv_val = u_int16_t(value);
```
Comment (nkaralis) [inline src/rdf4cpp/rdf/datatypes/xsd/UnsignedByte.hpp]: uint16_t

File: src/rdf4cpp/rdf/datatypes/xsd/NonNegativeInteger.hpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+ * Name of the datatype. This is kept so that we won't need to type it over and over again.
+ */
+constexpr static registry::ConstexprString xsd_nonNegativeInteger{"http://www.w3.org/2001/XMLSchema#nonNegativeInteger"};
+
+/**
+ * Defines the mapping between the LiteralDatatype IRI and the C++ datatype.
+ */
+template<>
+struct DatatypeMapping<xsd_nonNegativeInteger> {
+    using cpp_datatype = int64_t;
```
Comment (nkaralis) [inline src/rdf4cpp/rdf/datatypes/xsd/NonNegativeInteger.hpp]: Shouldn't this be uint64_t?

File: src/rdf4cpp/rdf/datatypes/xsd/PositiveInteger.hpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+ * Name of the datatype. This is kept so that we won't need to type it over and over again.
+ */
+constexpr static registry::ConstexprString xsd_positiveInteger{"http://www.w3.org/2001/XMLSchema#positiveInteger"};
+
+/**
+ * Defines the mapping between the LiteralDatatype IRI and the C++ datatype.
+ */
+template<>
+struct DatatypeMapping<xsd_positiveInteger> {
+    using cpp_datatype = int64_t;
```
Comment (nkaralis) [inline src/rdf4cpp/rdf/datatypes/xsd/PositiveInteger.hpp]: Shouldn't this be uint64_t?

File: src/rdf4cpp/rdf/datatypes/xsd/UnsignedInt.hpp [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+ * Name of the datatype. This is kept so that we won't need to type it over and over again.
+ */
+constexpr static registry::ConstexprString xsd_unsignedInt{"http://www.w3.org/2001/XMLSchema#unsignedInt"};
+
+/**
+ * Defines the mapping between the LiteralDatatype IRI and the C++ datatype.
+ */
+template<>
+struct DatatypeMapping<xsd_unsignedInt> {
+    using cpp_datatype = u_int32_t;
```
Comment (nkaralis) [inline src/rdf4cpp/rdf/datatypes/xsd/UnsignedInt.hpp]: uint32_t

File: src/rdf4cpp/rdf/datatypes/xsd/UnsignedLong.hpp [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+ * Name of the datatype. This is kept so that we won't need to type it over and over again.
+ */
+constexpr static registry::ConstexprString xsd_unsignedLong{"http://www.w3.org/2001/XMLSchema#unsignedLong"};
+
+/**
+ * Defines the mapping between the LiteralDatatype IRI and the C++ datatype.
+ */
+template<>
+struct DatatypeMapping<xsd_unsignedLong> {
+    using cpp_datatype = u_int64_t;
```
Comment (nkaralis) [inline src/rdf4cpp/rdf/datatypes/xsd/UnsignedLong.hpp]: uint64_t

File: src/rdf4cpp/rdf/datatypes/xsd/UnsignedLong.hpp [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+ * Specialisation of from_string template function.
+ */
+template<>
+inline LiteralDatatypeImpl<xsd_unsignedLong>::cpp_type LiteralDatatypeImpl<xsd_unsignedLong>::from_string(std::string_view s) {
+
+    const std::regex long_regex("[\\+]?[0-9]+");
+
+    if (std::regex_match(s.data(), long_regex)) {
+        auto ulong_val = std::strtoul(s.data(), nullptr, 10);
+        if (ulong_val < std::numeric_limits<u_int64_t>::min() || ulong_val > std::numeric_limits<u_int64_t>::max()) throw std::runtime_error("XSD Parsing Error");
```
Comment (nkaralis) [inline src/rdf4cpp/rdf/datatypes/xsd/UnsignedLong.hpp]: uint64_t

File: src/rdf4cpp/rdf/datatypes/xsd/UnsignedShort.hpp [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+ * Name of the datatype. This is kept so that we won't need to type it over and over again.
+ */
+constexpr static registry::ConstexprString xsd_unsignedShort{"http://www.w3.org/2001/XMLSchema#unsignedShort"};
+
+/**
+ * Defines the mapping between the LiteralDatatype IRI and the C++ datatype.
+ */
+template<>
+struct DatatypeMapping<xsd_unsignedShort> {
+    using cpp_datatype = u_int16_t;
```
Comment (nkaralis) [inline src/rdf4cpp/rdf/datatypes/xsd/UnsignedShort.hpp]: uint16_t

File: tests/datatype/tests_UnsignedByte.cpp [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    CHECK(std::string(datatypes::xsd::UnsignedByte::identifier) == correct_type_iri_cstr);
+
+    auto type_iri = IRI(datatypes::xsd::UnsignedByte::identifier);
+    CHECK(type_iri.identifier() == correct_type_iri_cstr);
+
+    using type = datatypes::xsd::UnsignedByte::cpp_type;
+
+    CHECK(std::is_same_v<type, u_int8_t>);
+
+    u_int8_t value = 1;
```
Comment (nkaralis) [inline tests/datatype/tests_UnsignedByte.cpp]: uint8_t

File: tests/datatype/tests_UnsignedInt.cpp [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    CHECK(std::string(datatypes::xsd::UnsignedInt::identifier) == correct_type_iri_cstr);
+
+    auto type_iri = IRI(datatypes::xsd::UnsignedInt::identifier);
+    CHECK(type_iri.identifier() == correct_type_iri_cstr);
+
+    using type = datatypes::xsd::UnsignedInt::cpp_type;
+
+    CHECK(std::is_same_v<type, u_int32_t>);
+
+    u_int32_t value = 1;
```
Comment (nkaralis) [inline tests/datatype/tests_UnsignedInt.cpp]: uint32_t

File: tests/datatype/tests_UnsignedShort.cpp [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    constexpr auto correct_type_iri_cstr = "http://www.w3.org/2001/XMLSchema#unsignedShort";
+
+    CHECK(std::string(datatypes::xsd::UnsignedShort::identifier) == correct_type_iri_cstr);
+
+    auto type_iri = IRI(datatypes::xsd::UnsignedShort::identifier);
+    CHECK(type_iri.identifier() == correct_type_iri_cstr);
+
+    using type = datatypes::xsd::UnsignedShort::cpp_type;
+
+    CHECK(std::is_same_v<type, u_int16_t>);
```
Comment (nkaralis) [inline tests/datatype/tests_UnsignedShort.cpp]: uint16_t

File: tests/datatype/tests_UnsignedLong.cpp [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    CHECK(std::string(datatypes::xsd::UnsignedLong::identifier) == correct_type_iri_cstr);
+
+    auto type_iri = IRI(datatypes::xsd::UnsignedLong::identifier);
+    CHECK(type_iri.identifier() == correct_type_iri_cstr);
+
+    using type = datatypes::xsd::UnsignedLong::cpp_type;
+
+    CHECK(std::is_same_v<type, u_int64_t>);
+
+    u_int64_t value = 1;
```
Comment (nkaralis) [inline tests/datatype/tests_UnsignedLong.cpp]: uint64_t

File: tests/datatype/tests_UnsignedLong.cpp [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    auto lit1 = Literal::make<datatypes::xsd::UnsignedLong>(value);
+    CHECK(lit1.value<datatypes::xsd::UnsignedLong>() == value);
+    CHECK(lit1.lexical_form() == std::to_string(value));
+
+    value = 0;
+    auto lit2 = Literal::make<datatypes::xsd::UnsignedLong>(value);
+    CHECK(lit2.value<datatypes::xsd::UnsignedLong>() == value);
+    CHECK(lit2.lexical_form() == std::to_string(value));
+
+    value = std::numeric_limits<u_int64_t>::max();
```
Comment (nkaralis) [inline tests/datatype/tests_UnsignedLong.cpp]: uint64_t

File: tests/datatype/tests_Int.cpp [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+    CHECK(std::string(datatypes::xsd::Int::identifier) == correct_type_iri_cstr);
+
+    auto type_iri = IRI(datatypes::xsd::Int::identifier);
+    CHECK(type_iri.identifier() == correct_type_iri_cstr);
+
+    using type = datatypes::xsd::Int::cpp_type;
+
+    CHECK(std::is_same_v<type, int32_t>);
+
+    auto value = 1;
```
Comment (nkaralis) [inline tests/datatype/tests_Int.cpp]: In some tests you define value using the keyword `auto`, whereas in other test cases you use the `cpp_type` of the datatype. Is there a reason for this?

**Review by nkaralis — CHANGES_REQUESTED** [review]

**Review by nkaralis — APPROVED** [review]

**Pull request closed without merging**